File: packages/sdk-typescript/src/x402.ts
/**
 * x402 helper for TypeScript / Node agents.
 *
 * Lets an agent hire a provider on Agora with a single call. Uses viem
 * under the hood for the on-chain bits.
 *
 *   import { hireWithX402 } from "@agora/sdk";
 *   const job = await hireWithX402({
 *     baseUrl: "https://api.agoraproto.org",
 *     requesterDid: me.did,
 *     providerDid: "did:agora:abc...",
 *     task: { prompt: "translate" },
 *     budgetUsdc: "2.50",
 *     rpcUrl: "https://sepolia.base.org",
 *     privateKey: "0x...",
 *     chain: "baseSepolia",
 *   });
 */

import ky from "ky";

const ESCROW_ABI = [
  {
    type: "function",
    name: "createJob",
    stateMutability: "nonpayable",
    inputs: [
      { name: "payee", type: "address" },
      { name: "amount", type: "uint256" },
      { name: "taskHash", type: "bytes32" },
      { name: "deadline", type: "uint64" },
    ],
    outputs: [{ name: "jobId", type: "uint256" }],
  },
  {
    type: "function",
    name: "submitResult",
    stateMutability: "nonpayable",
    inputs: [
      { name: "jobId", type: "uint256" },
      { name: "resultHash", type: "bytes32" },
    ],
    outputs: [],
  },
  {
    type: "function",
    name: "approveAndPay",
    stateMutability: "nonpayable",
    inputs: [{ name: "jobId", type: "uint256" }],
    outputs: [],
  },
  {
    type: "function",
    name: "refund",
    stateMutability: "nonpayable",
    inputs: [{ name: "jobId", type: "uint256" }],
    outputs: [],
  },
] as const;

const ERC20_ABI = [
  {
    type: "function",
    name: "approve",
    stateMutability: "nonpayable",
    inputs: [
      { name: "spender", type: "address" },
      { name: "amount", type: "uint256" },
    ],
    outputs: [{ name: "", type: "bool" }],
  },
  {
    type: "function",
    name: "allowance",
    stateMutability: "view",
    inputs: [
      { name: "owner", type: "address" },
      { name: "spender", type: "address" },
    ],
    outputs: [{ name: "", type: "uint256" }],
  },
] as const;

export interface PaymentRequired {
  version: string;
  chain: string;
  chain_id: number;
  asset: { kind: string; address: `0x${string}`; symbol: string; decimals: number };
  amount: string;
  fee_estimate: string;
  recipient_contract: `0x${string}`;
  function: string;
  args: {
    payee: `0x${string}`;
    amount: string;
    taskHash: `0x${string}`;
    deadline: number;
  };
  retry_header: string;
  expires_in_seconds: number;
}

export interface QuoteArgs {
  baseUrl: string;
  providerDid: string;
  task: Record<string, unknown>;
  budgetUsdc: string;
}

export async function quote(args: QuoteArgs): Promise<Record<string, unknown>> {
  return (await ky
    .post(`${args.baseUrl.replace(/\/$/, "")}/v1/x402/quote`, {
      json: {
        provider_did: args.providerDid,
        task: args.task,
        budget_usdc: args.budgetUsdc,
      },
      timeout: 10_000,
    })
    .json()) as Record<string, unknown>;
}

export interface HireWithX402Args {
  baseUrl: string;
  requesterDid: string;
  providerDid: string;
  task: Record<string, unknown>;
  budgetUsdc: string;
  rpcUrl: string;
  privateKey: `0x${string}`;
  /** Chain object compatible with viem (e.g. baseSepolia). Required. */
  chain: unknown;
  /** Deadline offset in seconds. Default 24h. */
  deadlineSeconds?: number;
  timeoutMs?: number;
}

/**
 * End-to-end x402 hire: server-quote, on-chain approve+createJob, server-confirm.
 *
 * Returns the final job representation from the server on success.
 * Requires `viem` to be installed in the consuming project.
 */
export async function hireWithX402(
  args: HireWithX402Args,
): Promise<Record<string, unknown>> {
  // Dynamic imports so the SDK doesn't hard-require viem for agents
  // that never use the on-chain helper.
  const { createPublicClient, createWalletClient, http, privateKeyToAccount } =
    await loadViem();

  const base = args.baseUrl.replace(/\/$/, "");
  const deadline = Math.floor(Date.now() / 1000) + (args.deadlineSeconds ?? 24 * 3600);
  const body = {
    requester_did: args.requesterDid,
    provider_did: args.providerDid,
    task: args.task,
    budget_usdc: args.budgetUsdc,
    deadline_unix: deadline,
  };

  // Step 1: expect 402.
  const initial = await fetch(`${base}/v1/x402/jobs`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (initial.status !== 402) {
    throw new Error(`expected 402, got ${initial.status}: ${await initial.text()}`);
  }
  const required = JSON.parse(initial.headers.get("X-Payment-Required") ?? "{}") as PaymentRequired;

  // Step 2: on-chain approve + createJob.
  const account = privateKeyToAccount(args.privateKey);
  const wallet = createWalletClient({ account, chain: args.chain as any, transport: http(args.rpcUrl) });
  const pub = createPublicClient({ chain: args.chain as any, transport: http(args.rpcUrl) });
  const amount = BigInt(required.amount);

  const allowance = (await pub.readContract({
    address: required.asset.address,
    abi: ERC20_ABI,
    functionName: "allowance",
    args: [account.address, required.recipient_contract],
  })) as bigint;
  if (allowance < amount) {
    const txApprove = await wallet.writeContract({
      address: required.asset.address,
      abi: ERC20_ABI,
      functionName: "approve",
      args: [required.recipient_contract, amount],
    });
    await pub.waitForTransactionReceipt({ hash: txApprove });
  }

  const txCreate = await wallet.writeContract({
    address: required.recipient_contract,
    abi: ESCROW_ABI,
    functionName: "createJob",
    args: [
      required.args.payee,
      amount,
      required.args.taskHash,
      BigInt(required.args.deadline),
    ],
  });
  await pub.waitForTransactionReceipt({ hash: txCreate });

  // Step 3: retry with X-Payment-Tx.
  const confirmed = await fetch(`${base}/v1/x402/jobs`, {
    method: "POST",
    headers: { "Content-Type": "application/json", "X-Payment-Tx": txCreate },
    body: JSON.stringify(body),
  });
  if (!confirmed.ok) {
    throw new Error(`server rejected: ${confirmed.status} ${await confirmed.text()}`);
  }
  return (await confirmed.json()) as Record<string, unknown>;
}

async function loadViem(): Promise<{
  createPublicClient: any;
  createWalletClient: any;
  http: any;
  privateKeyToAccount: any;
}> {
  try {
    const viem = await import("viem");
    const accounts = await import("viem/accounts");
    return {
      createPublicClient: (viem as any).createPublicClient,
      createWalletClient: (viem as any).createWalletClient,
      http: (viem as any).http,
      privateKeyToAccount: (accounts as any).privateKeyToAccount,
    };
  } catch (e) {
    throw new Error(
      "x402 helper requires `viem`. Install with: npm install viem",
    );
  }
}


// ───────────────────────────────────────────────────────────────────
// Lifecycle helpers: result / approve / refund
// Each performs the same 3-step dance as hireWithX402:
//   1. POST {endpoint} → 402 with X-Payment-Required (or 200 if already done)
//   2. on-chain tx via viem
//   3. POST {endpoint} again with X-Payment-Tx header → 200 with updated job
// ───────────────────────────────────────────────────────────────────


export interface LifecycleArgs {
  baseUrl: string;
  jobId: string;
  rpcUrl: string;
  privateKey: `0x${string}`;
  /** Chain object compatible with viem (e.g. baseSepolia). Required. */
  chain: unknown;
  timeoutMs?: number;
}

export interface SubmitResultArgs extends LifecycleArgs {
  result: Record<string, unknown>;
}

async function lifecycleCall(
  args: {
    baseUrl: string;
    path: string;
    body: Record<string, unknown>;
    rpcUrl: string;
    privateKey: `0x${string}`;
    chain: unknown;
    contractFunction: string;
    extractArgs: (required: PaymentRequired) => unknown[];
  },
): Promise<Record<string, unknown>> {
  const { createPublicClient, createWalletClient, http, privateKeyToAccount } =
    await loadViem();
  const base = args.baseUrl.replace(/\/$/, "");

  // Step 1
  const initial = await fetch(`${base}${args.path}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(args.body),
  });
  if (initial.status === 200) {
    return (await initial.json()) as Record<string, unknown>;
  }
  if (initial.status !== 402) {
    throw new Error(`expected 402, got ${initial.status}: ${await initial.text()}`);
  }
  const required = JSON.parse(initial.headers.get("X-Payment-Required") ?? "{}") as PaymentRequired;
  if ((required as any).function !== args.contractFunction) {
    throw new Error(
      `server asked for ${(required as any).function}, expected ${args.contractFunction}`,
    );
  }

  // Step 2
  const account = privateKeyToAccount(args.privateKey);
  const wallet = createWalletClient({ account, chain: args.chain as any, transport: http(args.rpcUrl) });
  const pub = createPublicClient({ chain: args.chain as any, transport: http(args.rpcUrl) });
  const tx = await wallet.writeContract({
    address: required.recipient_contract,
    abi: ESCROW_ABI,
    functionName: args.contractFunction,
    args: args.extractArgs(required),
  });
  await pub.waitForTransactionReceipt({ hash: tx });

  // Step 3
  const confirmed = await fetch(`${base}${args.path}`, {
    method: "POST",
    headers: { "Content-Type": "application/json", "X-Payment-Tx": tx },
    body: JSON.stringify(args.body),
  });
  if (!confirmed.ok) {
    throw new Error(`server rejected: ${confirmed.status} ${await confirmed.text()}`);
  }
  return (await confirmed.json()) as Record<string, unknown>;
}

/**
 * Provider-side: submit a result for an on-chain job.
 * Calls AgoraEscrow.submitResult(jobId, resultHash).
 * Only the registered payee can call this (contract enforces NotPayee).
 */
export async function submitResultWithX402(
  args: SubmitResultArgs,
): Promise<Record<string, unknown>> {
  return lifecycleCall({
    baseUrl: args.baseUrl,
    path: `/v1/x402/jobs/${args.jobId}/result`,
    body: { result: args.result },
    rpcUrl: args.rpcUrl,
    privateKey: args.privateKey,
    chain: args.chain,
    contractFunction: "submitResult",
    extractArgs: (required) => [
      BigInt((required.args as any).jobId),
      (required.args as any).resultHash,
    ],
  });
}

/**
 * Requester-side: approve a submitted job and release escrow.
 * Calls AgoraEscrow.approveAndPay(jobId).
 * Only the original payer can call this (contract enforces NotPayer).
 */
export async function approveWithX402(
  args: LifecycleArgs,
): Promise<Record<string, unknown>> {
  return lifecycleCall({
    baseUrl: args.baseUrl,
    path: `/v1/x402/jobs/${args.jobId}/approve`,
    body: {},
    rpcUrl: args.rpcUrl,
    privateKey: args.privateKey,
    chain: args.chain,
    contractFunction: "approveAndPay",
    extractArgs: (required) => [BigInt((required.args as any).jobId)],
  });
}

/**
 * Refund an unfulfilled on-chain job (deadline must have passed).
 * Calls AgoraEscrow.refund(jobId). Tx will revert if the deadline has
 * not yet elapsed and the caller is not the contract owner.
 */
export async function refundWithX402(
  args: LifecycleArgs,
): Promise<Record<string, unknown>> {
  return lifecycleCall({
    baseUrl: args.baseUrl,
    path: `/v1/x402/jobs/${args.jobId}/refund`,
    body: {},
    rpcUrl: args.rpcUrl,
    privateKey: args.privateKey,
    chain: args.chain,
    contractFunction: "refund",
    extractArgs: (required) => [BigInt((required.args as any).jobId)],
  });
}

File: packages/sdk-python/src/agora_sdk/x402.py
"""x402 helper for Python agents.

Lets an agent hire a provider on Agora with a single call:

    from agora_sdk.x402 import hire_with_x402

    job = await hire_with_x402(
        base_url="https://api.agoraproto.org",
        requester_did=my_did,
        provider_did="did:agora:abc...",
        task={"prompt": "translate to French"},
        budget_usdc="2.50",
        rpc_url="https://sepolia.base.org",
        private_key=my_eth_private_key,
    )
    print(job["id"], job["escrow_tx_hash"])

The helper performs the full 4-step x402 flow:
  1. POST /v1/x402/jobs (no payment) -> 402 with X-Payment-Required
  2. ERC20.approve(USDC, escrow_contract, amount)
  3. AgoraEscrow.createJob(payee, amount, taskHash, deadline)
  4. POST /v1/x402/jobs with X-Payment-Tx: <createJob tx hash>

This file depends on `web3` and `eth-account`. They are optional extras:

    pip install "agora-sdk[onchain]"
"""

from __future__ import annotations

import json
import time
from decimal import Decimal
from typing import Any

import httpx


_ESCROW_ABI = [
    {
        "type": "function",
        "name": "createJob",
        "stateMutability": "nonpayable",
        "inputs": [
            {"name": "payee", "type": "address"},
            {"name": "amount", "type": "uint256"},
            {"name": "taskHash", "type": "bytes32"},
            {"name": "deadline", "type": "uint64"},
        ],
        "outputs": [{"name": "jobId", "type": "uint256"}],
    },
    {
        "type": "function",
        "name": "submitResult",
        "stateMutability": "nonpayable",
        "inputs": [
            {"name": "jobId", "type": "uint256"},
            {"name": "resultHash", "type": "bytes32"},
        ],
        "outputs": [],
    },
    {
        "type": "function",
        "name": "approveAndPay",
        "stateMutability": "nonpayable",
        "inputs": [{"name": "jobId", "type": "uint256"}],
        "outputs": [],
    },
    {
        "type": "function",
        "name": "refund",
        "stateMutability": "nonpayable",
        "inputs": [{"name": "jobId", "type": "uint256"}],
        "outputs": [],
    },
]

_ERC20_ABI = [
    {
        "type": "function",
        "name": "approve",
        "stateMutability": "nonpayable",
        "inputs": [
            {"name": "spender", "type": "address"},
            {"name": "amount", "type": "uint256"},
        ],
        "outputs": [{"name": "", "type": "bool"}],
    },
    {
        "type": "function",
        "name": "allowance",
        "stateMutability": "view",
        "inputs": [
            {"name": "owner", "type": "address"},
            {"name": "spender", "type": "address"},
        ],
        "outputs": [{"name": "", "type": "uint256"}],
    },
]


async def quote(
    base_url: str,
    *,
    provider_did: str,
    task: dict[str, Any],
    budget_usdc: str,
    timeout: float = 10.0,
) -> dict[str, Any]:
    """Get a price quote without committing to a job."""
    async with httpx.AsyncClient(timeout=timeout) as http:
        r = await http.post(
            f"{base_url.rstrip('/')}/v1/x402/quote",
            json={
                "provider_did": provider_did,
                "task": task,
                "budget_usdc": budget_usdc,
            },
        )
        r.raise_for_status()
        return r.json()


async def hire_with_x402(
    base_url: str,
    *,
    requester_did: str,
    provider_did: str,
    task: dict[str, Any],
    budget_usdc: str,
    rpc_url: str,
    private_key: str,
    deadline_seconds: int = 24 * 3600,
    timeout: float = 60.0,
) -> dict[str, Any]:
    """Hire a provider on Agora via the x402 protocol — one call, four steps.

    Returns the final job dict on success. Raises on payment or
    verification failure.
    """
    try:
        from web3 import Web3
    except ImportError as e:  # pragma: no cover
        raise RuntimeError(
            "agora-sdk x402 helper requires `web3` and `eth-account`. "
            "Install with: pip install 'agora-sdk[onchain]'"
        ) from e

    deadline = int(time.time()) + deadline_seconds
    base = base_url.rstrip("/")

    async with httpx.AsyncClient(timeout=timeout) as http:
        # ── Step 1: request → expect 402 ──────────────────────────
        body = {
            "requester_did": requester_did,
            "provider_did": provider_did,
            "task": task,
            "budget_usdc": budget_usdc,
            "deadline_unix": deadline,
        }
        r = await http.post(f"{base}/v1/x402/jobs", json=body)
        if r.status_code != 402:
            r.raise_for_status()
            raise RuntimeError(f"expected 402, got {r.status_code}: {r.text[:200]}")
        required = json.loads(r.headers["X-Payment-Required"])

        # ── Step 2: on-chain approve + createJob ──────────────────
        w3 = Web3(Web3.HTTPProvider(rpc_url, request_kwargs={"timeout": 30}))
        acct = w3.eth.account.from_key(private_key)
        usdc = w3.eth.contract(
            address=Web3.to_checksum_address(required["asset"]["address"]),
            abi=_ERC20_ABI,
        )
        escrow = w3.eth.contract(
            address=Web3.to_checksum_address(required["recipient_contract"]),
            abi=_ESCROW_ABI,
        )
        amount = int(required["amount"])
        task_hash = bytes.fromhex(required["args"]["taskHash"].removeprefix("0x"))
        payee = Web3.to_checksum_address(required["args"]["payee"])

        # 2a. Approve if needed.
        existing = usdc.functions.allowance(acct.address, escrow.address).call()
        if existing < amount:
            approve_tx = usdc.functions.approve(escrow.address, amount).build_transaction(
                {"from": acct.address, "nonce": w3.eth.get_transaction_count(acct.address)}
            )
            signed = acct.sign_transaction(approve_tx)
            tx_hash = w3.eth.send_raw_transaction(signed.raw_transaction)
            w3.eth.wait_for_transaction_receipt(tx_hash, timeout=120)

        # 2b. createJob.
        create_tx = escrow.functions.createJob(
            payee, amount, task_hash, required["args"]["deadline"]
        ).build_transaction(
            {"from": acct.address, "nonce": w3.eth.get_transaction_count(acct.address)}
        )
        signed = acct.sign_transaction(create_tx)
        create_hash = w3.eth.send_raw_transaction(signed.raw_transaction)
        receipt = w3.eth.wait_for_transaction_receipt(create_hash, timeout=180)
        if receipt["status"] != 1:
            raise RuntimeError(f"createJob reverted: {create_hash.hex()}")
        create_hash_hex = "0x" + create_hash.hex()

        # ── Step 3: retry with X-Payment-Tx ───────────────────────
        r2 = await http.post(
            f"{base}/v1/x402/jobs",
            json=body,
            headers={"X-Payment-Tx": create_hash_hex},
        )
        r2.raise_for_status()
        return r2.json()


# ─────────────────────────────────────────────────────────────────────
# Lifecycle helpers — mirror the create flow for result/approve/refund.
# Each one performs the same 3-step dance: request → on-chain tx → retry.
# ─────────────────────────────────────────────────────────────────────


def _load_web3() -> Any:
    try:
        from web3 import Web3
    except ImportError as e:  # pragma: no cover
        raise RuntimeError(
            "agora-sdk x402 helper requires `web3` and `eth-account`. "
            "Install with: pip install 'agora-sdk[onchain]'"
        ) from e
    return Web3


async def _x402_lifecycle_call(
    *,
    base_url: str,
    path: str,
    body: dict[str, Any],
    rpc_url: str,
    private_key: str,
    contract_function: str,
    extract_args: Any,  # callable: (parsed_X-Payment-Required) -> tuple[Contract args]
    timeout: float,
) -> dict[str, Any]:
    """Shared driver for result/approve/refund. Internal."""
    Web3 = _load_web3()
    base = base_url.rstrip("/")

    async with httpx.AsyncClient(timeout=timeout) as http:
        # Step 1: trigger 402.
        r = await http.post(f"{base}{path}", json=body)
        if r.status_code == 200:
            # idempotent — server already applied this state change.
            return r.json()
        if r.status_code != 402:
            r.raise_for_status()
            raise RuntimeError(f"expected 402, got {r.status_code}: {r.text[:200]}")
        required = json.loads(r.headers["X-Payment-Required"])
        assert required["function"] == contract_function, (
            f"server told us to call {required['function']}, expected {contract_function}"
        )

        # Step 2: on-chain call.
        w3 = Web3(Web3.HTTPProvider(rpc_url, request_kwargs={"timeout": 30}))
        acct = w3.eth.account.from_key(private_key)
        escrow = w3.eth.contract(
            address=Web3.to_checksum_address(required["recipient_contract"]),
            abi=_ESCROW_ABI,
        )
        fn_args = extract_args(required)
        fn = getattr(escrow.functions, contract_function)(*fn_args)
        tx = fn.build_transaction(
            {"from": acct.address, "nonce": w3.eth.get_transaction_count(acct.address)}
        )
        signed = acct.sign_transaction(tx)
        tx_hash = w3.eth.send_raw_transaction(signed.raw_transaction)
        receipt = w3.eth.wait_for_transaction_receipt(tx_hash, timeout=180)
        if receipt["status"] != 1:
            raise RuntimeError(f"{contract_function} reverted: 0x{tx_hash.hex()}")
        tx_hex = "0x" + tx_hash.hex()

        # Step 3: retry with proof.
        r2 = await http.post(f"{base}{path}", json=body, headers={"X-Payment-Tx": tx_hex})
        r2.raise_for_status()
        return r2.json()


async def submit_result_with_x402(
    base_url: str,
    *,
    job_id: str,
    result: dict[str, Any],
    rpc_url: str,
    private_key: str,
    timeout: float = 60.0,
) -> dict[str, Any]:
    """Provider-side: submit a result for an on-chain job.

    Performs 402 → AgoraEscrow.submitResult(jobId, resultHash) → retry.
    Only the registered payee (whose key you must hold here) can call
    submitResult; the contract enforces NotPayee otherwise.
    """
    return await _x402_lifecycle_call(
        base_url=base_url,
        path=f"/v1/x402/jobs/{job_id}/result",
        body={"result": result},
        rpc_url=rpc_url,
        private_key=private_key,
        contract_function="submitResult",
        extract_args=lambda required: (
            int(required["args"]["jobId"]),
            bytes.fromhex(required["args"]["resultHash"].removeprefix("0x")),
        ),
        timeout=timeout,
    )


async def approve_with_x402(
    base_url: str,
    *,
    job_id: str,
    rpc_url: str,
    private_key: str,
    timeout: float = 60.0,
) -> dict[str, Any]:
    """Requester-side: approve a submitted job and release escrow.

    Performs 402 → AgoraEscrow.approveAndPay(jobId) → retry. Only the
    original payer can call approveAndPay; the contract enforces NotPayer
    otherwise.
    """
    return await _x402_lifecycle_call(
        base_url=base_url,
        path=f"/v1/x402/jobs/{job_id}/approve",
        body={},
        rpc_url=rpc_url,
        private_key=private_key,
        contract_function="approveAndPay",
        extract_args=lambda required: (int(required["args"]["jobId"]),),
        timeout=timeout,
    )


async def refund_with_x402(
    base_url: str,
    *,
    job_id: str,
    rpc_url: str,
    private_key: str,
    timeout: float = 60.0,
) -> dict[str, Any]:
    """Requester-side: refund an unfulfilled job (deadline expired).

    Performs 402 → AgoraEscrow.refund(jobId) → retry. The on-chain call
    is only valid once `block.timestamp > deadline` (or if the caller is
    the contract owner); otherwise the tx will revert.
    """
    return await _x402_lifecycle_call(
        base_url=base_url,
        path=f"/v1/x402/jobs/{job_id}/refund",
        body={},
        rpc_url=rpc_url,
        private_key=private_key,
        contract_function="refund",
        extract_args=lambda required: (int(required["args"]["jobId"]),),
        timeout=timeout,
    )
